src/node/filesystem/binary.node.ts

```diff
@@ -0,0 +1,341 @@
+    this.stream.resume();
+
+    return new Promise((resolve, reject) => {
+      if (this.pendingReadResolve !== undefined) {
+        reject(new UnexpectedError('Waiting for data failed. Unable to resolve pending read'));
+      }
+      this.pendingReadResolve = resolve;
+    })
+  }
+
+  private notifyData() {
+    if (this.pendingReadResolve !== undefined) {
+      this.pendingReadResolve();
+    }
+
+    this.stream.pause();
+    this.pendingReadResolve = undefined;
+  }
+
+  public async read(size = 1): Promise<Buffer | undefined> {
+    while (this.buffer.length < size) {
+      if (this.ended) {
+        if (this.buffer.length > 0) {
+          // yield remaining data
+          break;
+        } else {
+          // signal EOF
+          return undefined;
+        }
+      }
+
+      await this.waitForData();
+    }
+
+    const chunk = this.buffer.subarray(0, size); // this might need to be a copy
+    this.buffer = this.buffer.subarray(size);
+
+    return chunk;
+  }
+}
+
+export class ReadableClone extends Readable {
+  constructor(private stream: Readable, options?: ReadableOptions) {
+    super(options);
+
+    this.stream.on('data', (chunk) => {
+      this.push(chunk);
+    });
+
+    this.stream.on('end', () => {
+      this.push(null);
+    });
+
+    this.stream.on('error', (error) => {
+      this.emit('error', error);
+    })
+  }
+
+  public override _read(): void {}
+}
+
+export class FileContainer implements IDataContainer, IBinaryFileMeta, IInitializable, IDestructible {
+  private handle: FileHandle | undefined;
+  private stream: Readable | undefined;
+  private streamReader: StreamReader | undefined;
+  public filesize = Infinity;
+  public filename: string | undefined;
+  public mimetype: string | undefined;
+
+  constructor(public path: string, options: { filename?: string, mimetype?: string } = {}) {
+    this.mimetype = options.mimetype;
+    this.filename = options.filename;
+  }
+
+  public async read(size?: number): Promise<Buffer | undefined> {
+    if (!this.streamReader) {
+      throw new UnexpectedError('File not initialized');
+    }
+
+    return await this.streamReader.read(size);
+  }
+
+  public async initialize(): Promise<void> {
+    if (this.handle === undefined) {
+      try {
+        this.handle = await open(this.path, 'r');
+        const { size } = await this.handle.stat();
+        this.filesize = size;
+      } catch (error) {
+        throw handleNodeError(error);
+      }
+    }
+
+    if (this.handle === undefined) {
+      throw new UnexpectedError('Unable to initialize file');
+    }
+
+    let stream: Readable;
+    // We need to create a stream the old fashioned way for Node < 16.11.0
+    if (typeof this.handle.createReadStream !== 'function') {
+      stream = createReadStream(this.path);
+    } else {
+      stream = this.handle.createReadStream();
+    }
+
+    this.stream = new ReadableClone(stream);
+    this.streamReader = new StreamReader(new ReadableClone(stream));
```

Comment:
question: I don't quite understand the point of this cloning. What is it good for?

Reply:
All streams handling was changes. Resolving as outdated.